feat(artifacts): Add ArtifactStore to Kork (#1069)
spinnaker/kork: kork-artifacts/src/main/java/com/netflix/spinnaker/kork/artifacts/README.md

                                 |           |
                                 |   Orca    |
                                 |           |
                                 +-----------+
                                   |        ^
                                   |        | (outgoing artifact compressed)
                                   |        +----------------------+
                                   | (bake request)                |
                                   +---------------------+         |
                                                         v         |
        +---------------+            (fetch)           +-------------+
        |               |<-----------------------------|             |
        |  Clouddriver  |                              |  Rosco      |
        |               |                              |             |
        |  s3 get       |   (full artifact returned)   |  s3 stores  |
        |  artifacts    |----------------------------->|  artifacts  |
        +---------------+                              +-------------+

Artifact storage is divided into two operations of get and store, and there are

> primarily two services that utilizes each of these operations. Further the

Reviewer: ```suggestion
primarily two services that utilize each of these operations. Further the
```

artifact storage system relies on Spring's (de)serializing to call these
operations to limit the amount of code changes needed within these services.

When bootstrapping Spring we add in custom bean serializers and deserializers to
handle storage or retrieval of an artifact.

> Rosco is primarily used for baking artifacts which will generate something
> deployable. When Spring sends a request back to Orca for a bake request, the
> custom serializer injected at Orca startup, will use the artifact storage and
> store the artifact, and instead return a `remote/base64` artifact instead of the
> usual `embedded/base64`.

Reviewer: ```suggestion
Rosco is primarily used for baking artifacts which will generate something
deployable.  When Rosco responds to a bake request, the custom serializer
injected in Rosco at startup stores the artifact and returns a `remote/base64`
artifact instead of the usual `embedded/base64`.
```

# Artifact Storage

Artifact Storage is a feature that allows for embedded aritfacts to be
persisted to some storage, eg the S3ArtifactStore. Spinnaker keeps a history,
which is called the pipeline context, that contains everything that an
execution had done. Pipeline contexts can be very large, especially with
complicated pipeines, and size can be further increased when users have large
artifacts. Spinnaker will duplicate
these artifacts whenever any stage uses any of those artifacts. Using an
artifact store reduces this overhead by providing a reference link of,

> `ref://<spinnaker-application/<content-hash>`. This reduces the context size

Reviewer: ```suggestion
`ref://<spinnaker-application>/<content-hash>`. This reduces the context size
```

tremendously, but will vary depending on the size of the pipeline, as well as
how that artifact is used, but we've seen improvements of 80% for some
pipelines.

## Configuration

> To enable artifact storage, simple add this to your `spinnaker-local.yml` file

Reviewer: Please add some details about which microservices read from / write to s3.

```yaml

> artifactStore:

Reviewer 2: Seems like there is some discrepancy between `artifactStore` and `artifact-store`.
```suggestion
artifact-store:
```

  enabled: true
  s3:
    enabled: true
    bucket: spinnaker-qa-artifact-store-bucket
```

## Storage Options

### S3

[S3](https://aws.amazon.com/s3/) is an object store provided by AWS. The

> currently S3ArtifactStore implementation provides various ways to authenticate

Reviewer: ```suggestion
current S3ArtifactStore implementation provides various ways to authenticate
```

against AWS.

```yaml

> artifactStore:

Reviewer 2: ```suggestion
artifact-store:
```

  enabled: true
  s3:
    enabled: true
    profile: dev # if you want to authenticate using a certain profile
    region: us-west-2 # allows for specified regions
    bucket: spinnaker-qa-artifact-store-bucket
```

While the implementation is S3 specific, this does not limit usages of other S3
compatible storage engines. For example, something like

> [SeaweedFS](https://github.com/seaweedfs/seaweedfs) can be used to test locally

Reviewer: ```suggestion
[SeaweedFS](https://github.com/seaweedfs/seaweedfs) can be used to test locally.
```
and strike the next line.

with

## Local Testing

To test the artifact store locally, we will use SeaweedFS. To start the storage simply run
`docker run -p 8333:8333 chrislusf/seaweedfs server -s3`

Next enable the configuration

```yaml

> artifactStore:

Reviewer 2: ```suggestion
artifact-store:
```
